src/rgw/rgw_op.cc

```diff
@@ -6397,10 +6397,40 @@ void RGWCompleteMultipart::execute(optional_yield y)
     return;
   }
 
-  op_ret = upload->complete(this, y, s->cct, parts->parts, remove_objs, accounted_size, compressed, cs_info, ofs, s->req_id, s->owner, olh_epoch, s->object.get());
-  if (op_ret < 0) {
-    ldpp_dout(this, 0) << "ERROR: upload complete failed ret=" << op_ret << dendl;
-    return;
+  while (true) {
+    RGWObjVersionTracker objv_tracker = meta_obj->get_version_tracker();
+
+    op_ret = upload->complete(this, y, s->cct, parts->parts, remove_objs, accounted_size, compressed, cs_info, ofs, s->req_id, s->owner, olh_epoch, s->object.get());
```

Comment:
> do we really need to call `upload->complete()` multiple times? that writes the head object and updates the bucket index each time
> 
> cc @mattbenjamin i'm not familiar enough with the multipart re-upload fixes to understand the interaction here
> 
> can we just call `upload->complete()` once, then handle racing part uploads by re-listing the parts and deleting the ones that we'd otherwise orphan?

Certainly it's not a good idea to write the head object and updates the bucket index multiple times. Let me see how I can factor out the listing-parts and deleting.

Reply:
I'm not really either, that said, I'm not sure how problematic retrying this is practice.  I'm a bit worried about having to rebase checksum fixes onto some new behavior here.

Reply:
when `upload->complete()` calls `obj_op.write_meta()` to write the head object and update the bucket index, that also handles overwrites of existing objects by scheduling their tail objects for garbage collection. calling `upload->complete()` multiple times would look like overwrites, so we'd gc the part/tail objects that we want in the final upload

Reply:
i revisited the multipart reupload changes from https://github.com/ceph/ceph/pull/49709. that added a call to  `cleanup_part_history()` for each part in `RadosMultipartUpload::complete()`, which schedules all of that part's `past_prefixes` for gc. if we're adding new logic to clean up parts that were reuploaded after `complete()`, we should consult their `past_prefixes` too. this probably means tracking the set of prefixes we've already seen so we can tell which still need cleanup

p.s. that pr also handled cleanup in `RadosMultipartUpload::abort()`. the races that you're addressing for CompleteMultipartUpload will probably show up in AbortMultipartUpload too; should that be in scope for this pr?

Reply:
- added `cleanup_orphaned_parts` to cleaned up parts that we haven't cleaned up yet instead of calling `upload->complete()` multiple times
- added cleanup logic in `RadosMultipartUpload::abort()` too